RFC/RFC-9-on-chain-validators.md

```diff
@@ -0,0 +1,261 @@
+
+The final goal is to have a set of different upgradable smart contracts managed by the DAO needed that emulates the subgraphs.
+
+This document also describes a migration approach needed to make profile validations possible.
+
+## Proposed solution
+
+Subgraphs are being used by the content server to get:
+
+- Scenes
+- Collections items
+- ThirdParty Wearables
+- Ethereum & Polygon block numbers
+- Profiles
+
+The following sections will describe how to use only RPC calls and smart contracts to validate deployments by entity type.
+
+### Ethereum & Polygon block numbers
+
+Each entity deployment has a timestamp. We get the block number for that timestamp by using [subgraphs](https://github.com/decentraland/ethereum-blocks) per each chain. It is not possible to get a block number by timestamp directly with only one RPC request because timestamps are just part of the block metadata. In order to remove this dependency, we need extra work:
+
+- A) Use a [binary search](https://github.com/nachomazzara/eth-block-timestamp) or any other new way to get the closest block for a given timestamp.
+- B) Start adding block numbers to the deployments. For old deployments, we can create an immutable file with all the blocks per each timestamp used for previous deployments. This file will have a fixed size because new deployments will have the block number directly inside the deployment.
+
+### Profiles
+
+- **Entity**: Profile
+- **Pointer**: Ethereum Address. E.g: `0x5e5d9d1dfd87e9b8b069b8e5d708db92be5ade99`
+- **Contracts**: Decentraland Collections (Ethereum & Polygon). Decentraland Name.
+- **Check**: Every profile deployment checks if the address owns the wearables, emotes and Decentaland name.
+
+#### Decentraland Name
+
+The [DCLRegistrar](https://etherscan.io/address/0x2a187453064356c898cae034eaed119e1663acb8) contract has a method to get who is the owner of a specific domain: _`getOwnerOf(subdomain: string)`_.
+
+#### Wearables & Emotes
+
+Each entity has a `wearables` and `emotes` in the property `entity->metadata->avatars`. Wearables are being stored with an array format like `[wearable_urn_1, wearable_item_urn_2, ..., wearable_urn_N]` and Emotes as an array as well but with objects `[{slot: 1, urn: wave}, {slot: 2, urn: emote_item_urn_1}]`.
+Profiles are stored with the item's urn instead of the NFT itself urn. E.g: `urn:decentraland:matic:collections-v2:0x875146d1d26e91c80f25f5966a84b098d3db1fc8:1` (item); `urn:decentraland:matic:collections-v2:0x875146d1d26e91c80f25f5966a84b098d3db1fc8:1:1` (token id/NFT). Checking if a user has a specific item doesn't scale. The ERC721 standard has a method `ownerOf(tokenId: uint256)` to check who is the owner of the NFT. All the profiles are using only item ids instead of the specific NFT. _Remember that NFTs are items kind for Decentraland Collections_. To perform on-chain checks we should perform some changes to how we save profiles and how to validate old profile deployments.
```

Comment:
Profile `emotes` and `wearables` are stored ... ?

Reply:
@pedrotambo and @agusaldasoro. In the profiles, we are storing the item's urn not the NFT urn. The suggestion is not valid. How can I help to remove the confusion?

Reply:
Each `Profile` entity stores the `wearables` and `emotes` equipped by the user. Wearables are stored under `entity.metadata.avatar.wearables` as an array of URNs, and Emotes are stored under `entity.metadata.avatar.emotes` as an array of `{ slot: string, urn: string }`. In both cases the URNs used are the **item's** urn, not the **token's** urn. Checking if a user owns a specific item does not scale, since it would require looping within a contract method (users that own too many nfts or collections that are too big could cause this to lag or timeout) or making several RPC calls between the catalyst and the ethereum node (which also would not scale well). We should start storing the **token's** urn within the profiles to make the access checks possible in a single RPC call.